h/admin/views/badge.py

```diff
@@ -1,40 +1,53 @@
 # -*- coding: utf-8 -*-
 
+from pyramid import httpexceptions
 from pyramid.view import view_config
+from sqlalchemy.exc import IntegrityError
 
 from h import models
+from h.i18n import TranslationString as _
 
 
 @view_config(route_name='admin_badge',
              request_method='GET',
              renderer='h:templates/admin/badge.html.jinja2',
              permission='admin_badge')
-def badge_index(_):
-    return {"uris": models.Blocklist.all()}
+def badge_index(request):
+    return {"uris": request.db.query(models.Blocklist).all()}
 
 
 @view_config(route_name='admin_badge',
              request_method='POST',
              request_param='add',
-             renderer='h:templates/admin/badge.html.jinja2',
              permission='admin_badge')
 def badge_add(request):
+    uri = request.params['add']
+    item = models.Blocklist(uri=uri)
+    request.db.add(item)
+
+    # There's a uniqueness constraint on `uri`, so we flush the session,
+    # catching any IntegrityError and responding appropriately.
     try:
-        request.db.add(models.Blocklist(uri=request.params['add']))
-    except ValueError as exc:
-        request.session.flash(exc.message, 'error')
-    return badge_index(request)
+        request.db.flush()
+    except IntegrityError:
+        request.db.rollback()
```

Comment:
It's not even strictly necessary (the transaction is already in an aborted state), so probably not.